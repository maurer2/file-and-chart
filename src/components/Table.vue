<template>
  <table class="table">
    <thead>
      <tr>
        <template v-for="tableColumn in tableColumns">
          <th :key="tableColumn">
            {{ tableColumn | capitalizeFirstLetter }}
          </th>
        </template>
      </tr>
    </thead>
    <tbody>
      <tr v-for="row in numberOfRows" :key="'row-' + row">
        <template v-for="tableColumn in tableColumns">
          <td :key="tableColumn">
            {{ tableData[row - 1][tableColumn] }}
          </td>
        </template>
      </tr>
    </tbody>
  </table>
</template>

<script>
import flatten from 'lodash/flatten';

export default {
  name: 'Table',
  filters: {
    capitalizeFirstLetter(newString) {
      if (newString === undefined || newString.length === 0) {
        return '';
      }

      return `${newString.charAt(0).toUpperCase()}${newString.slice(1)}`;
    },
  },
  props: {
    tableData: {
      type: Array,
      default: () => [],
    },
  },
  computed: {
    numberOfRows() {
      return this.tableData.length;
    },
    /*
    numberOfColumns() {
      const fieldsEachEntry = this.tableData.map((tableRow) => Object.keys(tableRow).length);
      const largestNumberOfEntries = Math.max.apply(null, fieldsEachEntry);

      return largestNumberOfEntries;
    },
    */
    tableColumns() {
      const keysUniqueNested = this.tableData.map(tableRow => Object.keys(tableRow));
      const keysUniqueFlat = flatten(keysUniqueNested);
      const keysUnique = keysUniqueFlat.filter((key, index, keysWithDuplicates) => keysWithDuplicates.indexOf(key) === index);

      return keysUnique;
    },
  },
};
</script>

<style scoped lang="scss">
.table {
  width: 100%;
  table-layout: fixed;
  border-collapse: collapse;
  border: 1px solid black;

  thead tr {
    &:nth-child(odd) {
      background: whitesmoke;
    }
  }

  tbody tr {
    &:nth-child(even) {
      background: whitesmoke;
    }
  }

  th,
  td {
    padding: 0.5rem;
    border-left: 1px solid black;
  }

  td {
    border-top: 1px solid black;
  }
}
</style>
