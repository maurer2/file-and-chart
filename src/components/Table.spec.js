import { shallowMount } from '@vue/test-utils';
import Table from '@/components/Table.vue';

const tableData = [
  { name: 'Bob', age: '15' },
  { name: 'Boots', age: '21' },
  { name: 'Cat', age: '32' },
  { name: 'Felix', age: '44' },
  { name: 'Garfield', age: '51' },
];

describe('Table', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Table, {
      propsData: {
        tableData,
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Table snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Table contains all elements and child elements', () => {
    expect(wrapper.find('.table').exists()).toBe(true);
  });

  test('Table has head and body', () => {
    expect(wrapper.find('thead').exists()).toBe(true);
    expect(wrapper.find('tbody').exists()).toBe(true);
  });

  test('Table has correct number of rows', () => {
    expect(wrapper.findAll('thead tr').length).toBe(1);
    expect(wrapper.findAll('tbody tr').length).toBe(tableData.length);
  });

  test('Getter numberOfRows returns correct number of rows', () => {
    expect(wrapper.vm.numberOfRows).toBe(tableData.length);

    wrapper.setProps({ tableData: [] });
    expect(wrapper.vm.numberOfRows).toBe(0);
  });

  test('Getter tableColumns returns array of strings', () => {
    expect(wrapper.vm.tableColumns).toEqual(['name', 'age']);

    wrapper.setProps({ tableData: [] });
    expect(wrapper.vm.tableColumns).toEqual([]);
  });
});
