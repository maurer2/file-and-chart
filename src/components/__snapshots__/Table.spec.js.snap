// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Table snapshot 1`] = `
<table class="table">
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Age
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Bob
      </td>
      <td>
        15
      </td>
    </tr>
    <tr>
      <td>
        Boots
      </td>
      <td>
        21
      </td>
    </tr>
    <tr>
      <td>
        Cat
      </td>
      <td>
        32
      </td>
    </tr>
    <tr>
      <td>
        Felix
      </td>
      <td>
        44
      </td>
    </tr>
    <tr>
      <td>
        Garfield
      </td>
      <td>
        51
      </td>
    </tr>
  </tbody>
</table>
`;
